Simplify page warmer stop, fix method name spelling error.

MishaDemianenko opened the pull request:
Use plain sync block in PAgeCacheWarmer stop.
Rename Files.createOrOpenAsOuputStream -> Files.createOrOpenAsOutputStream

MishaDemianenko committed:
Simplify page warmer stop, fix method name spelling error.

Use plain sync block in PAgeCacheWarmer stop.
Rename Files.createOrOpenAsOuputStream -> Files.createOrOpenAsOutputStream

klaren commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/PageCacheWarmer.java:
```diff
@@ -132,7 +126,7 @@ public void stop()
      * @throws IOException if anything goes wrong while reading the profiled data back in.
      */
-    public synchronized OptionalLong reheat() throws IOException
+    synchronized OptionalLong reheat() throws IOException
```
You should check `stopped` at the beginning of the method instead.

klaren replied:
Same for `profile()`

klaren reviewed (changes requested):
You should also get rid of the `stopped` checks inside the methods since they do nothing now.

This change makes it so the reheating task cannot be aborted, so I believe we have to interrupt them on `org.neo4j.kernel.impl.pagecache.PageCacheWarmerKernelExtension.java:164`.

MishaDemianenko commented:
@klaren you are right fixing...

MishaDemianenko committed:
Restartable page cache warmer.
Update synchronisation in stop method.
Avoid double registration of same file name provider in file name
neo store file listing.

MishaDemianenko commented:
@klaren i change things a bit and come back to synch block in a bit different fashion, all thread waiting, separate locks and so on looks like a more complex thing; 
Also added the possibility to restart since extensions are lifecycle instances by nature and can be restarted.

klaren commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/PageCacheWarmer.java:
```diff
@@ -113,16 +104,22 @@ public synchronized Resource addFilesTo( Collection<StoreFileMetadata> coll ) th
-            // This synchronised block means we'll wait for any reheat() or profile() to notice our raised stopped flag,
-            // before we return to the caller. This helps avoid races, e.g. if the page cache is closed while this page
-            // cache warmer is still holding on to some mapped pages.
+            executor.shutdown();
```
Still looks strange since your synchronizing on a thread-safe object. Someone coming here might believe that you can remove that block, completely missing the intended barrier this synchronize provide.

I would add an additional variable that indicates that the profile/reheat is running and wait for that to flip to make this intention clear.
```
        // wait for shutdown
        synchronized ( this ) {
            while ( working ) {
                wait();
            }
        }
```

klaren reviewed (changes requested):
There is no test trying to stop a long-running reheat. Should be fairly easy to mock a `PageFile` with a latch. Should we have that?

MishaDemianenko committed:
Renaming and extract stop of warmer resource into separate method

MishaDemianenko committed:
Start stop of page cache warmer. Track start of extension to be able to
determine if warmer also needs to be shutdown.
